src/components/token/TokenBalances.js

```diff
@@ -0,0 +1,327 @@
+import helpers from '../../utils/helpers';
+
+/**
+ * Displays custom tokens in a table with pagination buttons and a search bar.
+ */
+class TokenBalances extends React.Component {
+  /**
+   * Structure that contains the attributes that will be part of the page URL
+   */
+  pagination = new PaginationURL({
+    'sortBy': { required: false },
+    'order': { required: false },
+  });
+
+  constructor(props) {
+    super(props);
+
+    /**
+     * tokenBalances: List of token balances currently being rendered.
+     *         Each token balance element must have the fields: address, locked_balance, unlocked_balance, total, token_id and sort.
+     *         id, name, symbol are strings; nft is boolean; transaction_timestamp is long.
+     *         Sort is an array with two strings, The value is given by ElasticSearch and it is passed back when we want to change page
+     * hasAfter: Indicates if a next page exists
+     * hasBefore: Indicates if a previous page exists
+     * sortBy: Which field to sort (uid, name, symbol)
+     * order: If sorted field must be ordered asc or desc
+     * page: Current page. Used to know if there is a previous page
+     * pageSearchAfter: Calculates the searchAfter param that needs to be passed to explorer-service in order to get the next/previous page.
+     *                  We use this array to store already-calculated values,
+     *                  so we do not need to recalculate them if user is requesting an already-navigated page.
+     * loading: Initial loading, when user clicks on the Tokens navigation item
+     * isSearchLoading: Indicates if search results are being retrieved from explorer-service
+     * calculatingPage: Indicates if next page is being retrieved from explorer-service
+     * error: Indicates if an unexpected error happened when calling the explorer-service
+     * tokenBalanceInformationError: Indicates if an unexpected error happened when calling the token balance information service 
+     * maintenanceMode: Indicates if explorer-service or its downstream services are experiencing problems. If so, maintenance mode is enabled as
+     *                  a "circuit breaker" to remove additional load until the team fixes the problem
+     */
+    this.state = {
+      tokenId: '00',
+      tokenBalances: [],
+      hasAfter: false,
+      hasBefore: false,
+      sortBy: 'total',
+      order: 'desc',
+      page: 1,
+      pageSearchAfter: [],
+      loading: false,
+      isSearchLoading: false,
+      calculatingPage: false,
+      error: false,
+      tokenBalanceInformationError: false,
+      maintenanceMode: this.props.maintenanceMode,
+      transactionsCount: 0,
+      addressesCount: 0,
+    };
+  }
+
+  componentDidMount = async () => {
+    if (this.state.maintenanceMode) {
+      return;
+    }
+
+    // 'Click' on search to make the first query
+    const queryParams = this.pagination.obtainQueryParams();
+
+    await this.setState({
```

Comment:
I think `this.setState` is not awaitable, as I detailed in another comment.

Reply:
It is not well documented, but does returns a promise 

This code has a race condition without it

Reply:
Amd because of this, I think we are in danger of falling in this pitfall here as well:

> setState() does not always immediately update the component. It may batch or defer the update until later. This makes reading this.state right after calling setState() a potential pitfall.

https://reactjs.org/docs/react-component.html#setstate

We are setting the `sortBy` and `order` in the state, and we will need to read them inside `this.performSearch` which is called immediately after.

Maybe we should call `this.pagination.obtainQueryParams();` in the constructor instead and set the initial state there. (I think the query params are available at construction time)

Reply:
> It is not well documented, but does returns a promise
> 
> This code has a race condition without it

Sorry about the other response, I didn't see your comment before.

But I think that's weird. I went to check React's code, and it seems they don't return a Promise indeed: https://github.com/facebook/react/blob/main/packages/react/src/ReactBaseClasses.js#L56

It seems weird to me that this `await` is making any difference. But I believe when you say so, I'm just curious to how exactly this happens, and whether `setState` really returns a Promise somehow.

Reply:
Is it possible that something similar to this is happening in your case?
https://stackoverflow.com/a/47022453/3477266

Reply:
You are right, that stackoverflows explains why this is working

Refactored in c67608b9dd1fe17265aecc17822d0fe19e1af2bf